### DISCUSSION TITLE
SCIP Strong Branching Scores

### DISCUSSION THREAD

**Discussion by cwfparsonson** [ds4dm/ml4co-competition#62] [Q&A]
Hi,

How does SCIP calculate strong branching scores? Apologies if this is documented somewhere, but I haven't been able to find it in the SCIP documentation.

From the literature I thought strong branching was a heuristic which does a 1-step-lookahead to see the change in the dual bound resulting from branching on each variable, and then chooses the variable which leads to the greatest change in the dual bound. However, if you print the SB scores and the change in the dual bound, they do not match. In fact, it seems that sometimes variables with 0 change in the dual bound have the highest strong branching scores, even when other variables would have led to a non-zero dual bound change for the next step. 

I thought my branching environment was disabling SCIP heuristics, but perhaps the SB scores do more than a 1-step lookahead, or use some heuristic(s) to outperform 1-step-lookahead strong branching?

The below code is what I use to show that strong branching scores != change in dual bound. To show the interesting cases where variables with 0 dual bound change can have the highest strong branching scores, you'd need to run the below in a loop until you generate an instance which leads to this behaviour.

```python
import ecole

class StrongBranchingAgent:
    def __init__(self, name='sb'):
        self.name = name
        self.strong_branching_function = ecole.observation.StrongBranchingScores()

    def before_reset(self, model):
        self.strong_branching_function.before_reset(model)
    
    def extract(self, model, done, *args):
        return self.strong_branching_function.extract(model, done)

    def action_select(self, action_set, model, done):
        scores = self.extract(model, done)[action_set]
        action = scores.argmax()
        
        return action, scores
    
    
agent = StrongBranchingAgent()
env = EcoleBranching(observation_function='default',
                          information_function='default',
                          reward_function='default',
                          scip_params='default')
env.seed(0)
instances = ecole.instance.SetCoverGenerator(n_rows=100, n_cols=100, density=0.05)

num_episodes = 1
for ep in range(num_episodes):
    # find an instance not pre-solved by environment
    obs = None
    while obs is None:
        env.seed(0)
        instance = next(instances)
        instance_before_reset = instance.copy_orig()
        agent.before_reset(instance_before_reset.copy_orig())
        obs, action_set, reward, done, info = env.reset(instance)
    
    # solve instances
    t = 1
    while not done:
        m = env.model.as_pyscipopt()
        prev_dual = m.getDualbound()
        action_idx, scores = agent.action_select(action_set, env.model, done)
        action = action_set[action_idx]
        obs, action_set, reward, done, info = env.step(action)
        m = env.model.as_pyscipopt()
        curr_dual = m.getDualbound()
        print(f'Step {t} action {action} | SB score: {scores[action_idx]} | Change in dual bound: {curr_dual-prev_dual}')
        t += 1
```

Output:

```
Step 1 action 49 | SB score: 326.56387499999965 | Change in dual bound: 13.75
Step 2 action 3 | SB score: 100.00200000999949 | Change in dual bound: 11.0
```

For completeness, here is the branching environment I'm using:

```python
DEFAULT_SCIP_PARAMS = {'separating/maxrounds': 0,
                       'separating/maxroundsroot': 0,
                       'separating/maxcuts': 0,
                       'separating/maxcutsroot': 0,
                       'presolving/maxrounds': 0,
                       'presolving/maxrestarts': 0,
                       'propagating/maxrounds':0,
                       'propagating/maxroundsroot':0,
                       'lp/initalgorithm':'d',
                       'lp/resolvealgorithm':'d',
                       'limits/time': 3600}
class EcoleBranching(ecole.environment.Branching):
    def __init__(
        self,
        observation_function='default',
        information_function='default',
        reward_function='default',
        scip_params='default',
    ):
        '''
        Args:
            scip_params (dict, str): Custom dictionary or one of: 'default', 
                'ml4co_item_placement', 'ml4co_load_balancing', 'ml4co_anonymous'.
        '''
        # save string names so easy to initialise new environments
        if type(observation_function) == str:
            self.str_observation_function = observation_function
        else:
            self.str_observation_function = None
        if type(information_function) == str:
            self.str_information_function = information_function
        else:
            self.str_information_function = None
        if type(reward_function) == str:
            self.str_reward_function = reward_function
        else:
            self.str_reward_function = None
        if type(scip_params) == str:
            self.str_scip_params = scip_params
        else:
            self.str_scip_params = None

        # init reward functions from strings if needed
        if reward_function == 'default':
            reward_function = ({
                     'num_nodes': -ecole.reward.NNodes(),
                     'lp_iterations': -ecole.reward.LpIterations(),
                     'primal_integral': -ecole.reward.PrimalIntegral(),
                     'dual_integral': ecole.reward.DualIntegral(),
                     'primal_dual_integral': -ecole.reward.PrimalDualIntegral(),
                     'solving_time': -ecole.reward.SolvingTime(),
                 })
        if information_function == 'default':
            information_function=({
                     'num_nodes': ecole.reward.NNodes().cumsum(),
                     'lp_iterations': ecole.reward.LpIterations().cumsum(),
                     'solving_time': ecole.reward.SolvingTime().cumsum(),
                 })
        if observation_function == 'default':    
            observation_function=(
                     ecole.observation.NodeBipartite()
                 )
        if scip_params == 'default':
            scip_params = DEFAULT_SCIP_PARAMS
        
        super(EcoleBranching, self).__init__(
            observation_function=observation_function,
            information_function=information_function,
            reward_function=reward_function,
            scip_params=scip_params
        )
```

**Comment by ambros-gleixner** [comment]
Regarding SCIP, a good reference is Tobias Achterberg's thesis, https://opus4.kobv.de/opus4-zib/frontdoor/index/index/docId/1112. Note that there are two dual bound improvements per strong branching variable: for the down and for the up child. Both need to be combined using a scoring function, which can be different than the minimum. I cannot comment on the ecole output, though.

**Comment by gasse** [comment]
Hi @cwfparsonson,

I think there are a few issues with your experiment.

1) you should make sure that both your branching environment (`ecole.environment.Branching`) and your SB observation function (`ecole.observation.StrongBranchingScores`) are considering the same sets of branching candidates.
```python
ecole.observation.StrongBranchingScores(pseudo_candidates=True)
ecole.environment.Branching(pseudo_candidates=True)
```
or
```python
ecole.observation.StrongBranchingScores(pseudo_candidates=False)
ecole.environment.Branching(pseudo_candidates=False)
```

(By default those sets do not match unfortunately, which is something we'll fix in the next release of Ecole after the competition).

2) **Edit: this is wrong (see answer below).** ~~The way you pick the best SB variable and return its score does not seem right. Given that both your observation function and the environment consider the same set of branching candidates, instead of~~
```python
        scores = self.extract(model, done)[action_set]
        action = scores.argmax()
```
~~you should have~~
```python
        scores = self.extract(model, done)
        best_score = scores[scores.argmax()]
        best_action = action_set[scores.argmax()]
```

3) the way you implemented your agent to extract the SB scores from SCIP is not the intended usage with Ecole, so you may get an unexpected behavior. You seem to apply `before_reset` to a duplicate of the model in the environment.
```python
        instance = next(instances)
        instance_before_reset = instance.copy_orig()
        agent.before_reset(instance_before_reset.copy_orig())
        obs, action_set, reward, done, info = env.reset(instance)
```
I would suggest to extract the SB scores by using an observation function which is passed the environment when it is instantiated, and then letting the environment call the extraction function at the proper time.

4) as @ambros-gleixner pointed out, the SB score is not the minimum LP value as you described (which would match with a local dual bound improvement). By default in SCIP the SB scores consist in a multiplication of the gains (child LP value - root LP value) obtained in the children nodes after branching. The implementation can be seen [here](https://www.scipopt.org/doc/html/branch__vanillafullstrong_8c_source.php) at lines 202, 203, 232. In addition, the SB scores are local (parent node LP value vs children nodes LP value), while the dual bound is global (minimum LP value in the whole tree). So, unless you are at the root node, the value of the SB scores at a node can generally be expected to be unrelated to the value of the global dual bound.

I hope this helps,

Best,
Maxime

**Reply by cwfparsonson** [reply]
Many thanks for taking the time to write such a detailed reply @gasse. 


> 1. you should make sure that both your branching environment (`ecole.environment.Branching`) and your SB observation function (`ecole.observation.StrongBranchingScores`) are considering the same sets of branching candidates.
> 
> ```python
> ecole.observation.StrongBranchingScores(pseudo_candidates=True)
> ecole.environment.Branching(pseudo_candidates=True)
> ```
> 
> or
> 
> ```python
> ecole.observation.StrongBranchingScores(pseudo_candidates=False)
> ecole.environment.Branching(pseudo_candidates=False)
> ```
> 
> (By default those sets do not match unfortunately, which is something we'll fix in the next release of Ecole after the competition).

Thank you for pointing this out, I had not realised there was a difference between these defaults. Does this mean there is an error in the `ecole` [GitHub Branching example](https://github.com/ds4dm/ecole/blob/master/examples/branching-imitation.ipynb), which uses the default `pseudo_candidates` for both `Branching()` and `StrongBranchingScores()`, therefore there's a discrepancy between the actions produced by the environment and the strong branching agent?


> 1. The way you pick the best SB variable and return its score does not seem right. Given that both your observation function and the environment consider the same set of branching candidates, instead of
> 
> ```python
>         scores = self.extract(model, done)[action_set]
>         action = scores.argmax()
> ```
> 
> you should have
> 
> ```python
>         scores = self.extract(model, done)
>         best_score = scores[scores.argmax()]
>         best_action = action_set[scores.argmax()]
> ```

I'm not too sure what might be wrong with my implementation here. `scores = self.extract(model, done)[action_set]` extracts the scores and filters the invalid scores using `action_set` as a mask, and in this context `scores.argmax()` just returns the index of `action_set` (the valid actions) which had the highest score, so you can then do `action_set[action]` to get the raw action rather than the index. Unless I've missed something, I think this is the same as the [GitHub Branching example](https://github.com/ds4dm/ecole/blob/master/examples/branching-imitation.ipynb), which uses `action = action_set[scores[action_set].argmax()]`. With your implementation, `scores` in `best_score = scores[scores.argmax()]` would contain invalid actions, which sometimes have `nan` scores so wouldn't you need to handle taking `argmax()` of a `nan` array?



> 1. the way you implemented your agent to extract the SB scores from SCIP is not the intended usage with Ecole, so you may get an unexpected behavior. You seem to apply `before_reset` to a duplicate of the model in the environment.
> 
> ```python
>         instance = next(instances)
>         instance_before_reset = instance.copy_orig()
>         agent.before_reset(instance_before_reset.copy_orig())
>         obs, action_set, reward, done, info = env.reset(instance)
> ```
> 
> I would suggest to extract the SB scores by using an observation function which is passed the environment when it is instantiated, and then letting the environment call the extraction function at the proper time.

I understand that the intended usage is to return these scores as part of an observation, but do you think it would be acceptable to pass the original instance rather than a copy? I.e. to replace what you've quoted with:

```python
instance = next(instances)
instance_before_reset = instance.copy_orig()
agent.before_reset(instance)
obs, action_set, reward, done, info = env.reset(instance)
```

I find it useful to have `instance_before_reset` saved so I can instantiate separate environments from the same instance, but I don't see a need to pass this copy to the agent beforehand.




> 1. as @ambros-gleixner pointed out, the SB score is not the minimum LP value as you described (which could match with a local dual bound improvement). By default in SCIP the SB scores consist in a multiplication of the gains (child LP value - root LP value) obtained in the children nodes after branching. The implementation can be seen [here](https://www.scipopt.org/doc/html/branch__vanillafullstrong_8c_source.php) at lines 202, 203, 232. In addition, the SB scores are local (parent node LP value vs children nodes LP value), while the dual bound is global (minimum LP value in the whole tree). So, unless you are at the root node, the value of the SB scores at a node can generally be expected to be unrelated to the value of the global dual bound.

Thanks very much to you and @ambros-gleixner for helping me understand this!

**Comment by gasse** [comment]
Hi @cwfparsonson ,

Ah sorry my bad, comment 2) in my answer above was erroneous. Yes you're right, the SB scores observation (and also the pseudocost scores observation) in Ecole consists in an array the size of the number of variables in the LP. This array can directly be indexed using the `action_set` from the branching environment. For variables whose score is not available, the array is filled with `nan` values (as can be seen [here](https://github.com/ds4dm/ecole/blob/master/libecole/src/observation/strongbranchingscores.cpp#L71-L72)). Your code to pick the best SB-scoring variable therefore seems correct:
```python
scores = self.extract(model, done)[action_set]
action_idx = scores.argmax()
action = action_set[action_idx]
```

Regarding comment 1) I think you're right as well, there seems to be an inconsistency in the branching tutorial. We'll look into in. Thanks for pointing this out !

Last, about the observation extraction and the `before_reset` call. If you intend to use the same agent with different environments, you should definitely keep the data extraction part (`model->observation`) and the policy part (`observation+action_set->action`) separate. The observation functions in Ecole are meant to be stateful, and are supposed to be used for a single environment at a time, and are reset at the beginning of every episode (hence the `before_reset` method). For example, some observation functions need to register [`event handler`](https://github.com/ds4dm/ecole/blob/master/libecole/src/reward/boundintegral.cpp#L23) into the SCIP model to extract some data, or some others may rely on a [cache system](https://github.com/ds4dm/ecole/blob/master/libecole/src/observation/khalil-2016.cpp#L608) so that expensive computations are only made once at the root node. Your implementation might work at the moment, but to avoid any surprises later on I strongly suggest that you keep both concepts of observation extraction / decision making separate, as is done [here](https://github.com/ds4dm/ml4co-competition/blob/main/submissions/example/agents/dual.py) for example. (Usage [here](https://github.com/ds4dm/ml4co-competition/blob/main/common/evaluate.py#L94-L103)).

Anyway thanks for asking questions, it helps us know what we can improve on our side :)

Hope that helps,

Best,
Maxime
